hyp3lib/rtc2color.py

```diff
@@ -1,144 +1,147 @@
+    """
+
+    # Suppress GDAL warnings but raise python exceptions
+    # https://gis.stackexchange.com/a/91393
+    gdal.UseExceptions()
+    gdal.PushErrorHandler('CPLQuietErrorHandler')
+
+    clean_threshold = pow(10.0, -48.0 / 10.0)  # db to power
+    power_threshold = pow(10.0, threshold / 10.0)  # db to power
+
+    # used scale the results to fit inside RGB 1-255 (ints), with 0 for no/bad data
+    scale_factor = 1.0 if real else 254.0
+    out_type = gdal.GDT_Float32 if real else gdal.GDT_Byte
+
+    copol = gdal.Open(copol_tif)
+    crosspol = gdal.Open(crosspol_tif)
+
+    geotransform = copol.GetGeoTransform()
+    proj_wkt = copol.GetProjectionRef()
+
+    cols = min(copol.RasterXSize, crosspol.RasterXSize)
+    rows = min(copol.RasterYSize, crosspol.RasterYSize)
+
+    logging.info(f'Reading co-pol image {copol_tif}')
+    cp = np.nan_to_num(copol.GetRasterBand(1).ReadAsArray()[:rows, :cols])
+    copol = None  # because gdal is weird
+
+    if amp:
+        cp *= cp
+
+    if cleanup:
+        cp[cp < clean_threshold] = 0
+    else:
+        # FIXME: Should this be here or before amp conversion to power? That is,
+        #        are the negative amp values, and if so, are they indicative of
+        #        bad data? This *was* done before amp conversion before...
+        cp[cp < 0] = 0
+
+    # Read cross-pol image
+    logging.info(f'Reading cross-pol image {crosspol_tif}')
+    xp = np.nan_to_num(crosspol.GetRasterBand(1).ReadAsArray()[:rows, :cols])
+    crosspol = None  # because gdal is weird
+
+    if amp:
+        xp *= xp
+
+    if cleanup:
+        xp[xp < clean_threshold] = 0
+    else:
+        # FIXME: Should this be here or before amp conversion to power? That is,
+        #        are the negative amp values, and if so, are they indicative of
+        #        bad data? This *was* done before amp conversion before...
+        xp[xp < 0] = 0
+
+    # Find all our no data and bad data pixels
+    # NOTE: we're using crosspol here because it will typically have the most bad
+    # data and we want the same mask applied to all 3 channels (otherwise, we'll
+    # accidentally be changing colors from intended.
+    invalid_xp_mask = ~(xp > 0)
+    # mask for applying colors
+    below_threshold_mask = xp < power_threshold
+
+    driver = gdal.GetDriverByName('GTiff')
+    out_raster = driver.Create(out_tif, cols, rows, 3, out_type, ['COMPRESS=LZW'])
+    out_raster.SetGeoTransform((geotransform[0], geotransform[1], 0, geotransform[3], 0, geotransform[5]))
+    out_raster_srs = osr.SpatialReference()
+    out_raster_srs.ImportFromWkt(proj_wkt)
+    out_raster.SetProjection(out_raster_srs.ExportToWkt())
+
+    logging.info('Calculating color decomposition components')
+
+    zp = np.arctan(np.sqrt(np.clip(cp - xp, 0, None))) * 2.0 / np.pi
+    zp[~below_threshold_mask] = 0
+
+    logging.info('Calculate red channel and save in GeoTIFF')
+    rp = 2.0 * np.sqrt(np.clip(cp - 3.0 * xp, 0, None))
```

Comment:
handling of each band has the same structure except for this initial calculation; wondering if there's a way to abstract the math (via a lambda?) so each band can be it's own function call

Reply:
Because I don't have any reference for the math/theory behind these calculations, I've had a hard time deciding on the form the functional abstraction should take for these. They definitely could be pulled into a function though.